File: VideoCompress.java
// VideoCompress.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;

public class VideoCompress {
    private static final Set<String> INPUT_EXTS = new HashSet<>(Arrays.asList(".mp4", ".mkv", ".avi", ".mov", ".wmv", ".flv", ".webm", ".m4v", ".3gp", ".ogv"));
    private static final Map<String, String> VIDEO_CODECS = new HashMap<>();
    private static final Set<String> AUDIO_CODECS = new HashSet<>(Arrays.asList("aac", "mp3", "opus", "ac3"));
    private static final Set<String> PRESETS = new HashSet<>(Arrays.asList("ultrafast", "superfast", "veryfast", "faster", "fast", "medium", "slow", "slower", "veryslow"));

    static {
        VIDEO_CODECS.put("h264", "libx264");
        VIDEO_CODECS.put("hevc", "libx265");
        VIDEO_CODECS.put("vp9", "libvpx-vp9");
        VIDEO_CODECS.put("av1", "libaom-av1");
    }

    public static void checkFFmpeg() {
        try {
            Process p = Runtime.getRuntime().exec(new String[]{"ffmpeg", "-version"});
            if (p.waitFor() != 0) {
                throw new Exception();
            }
        } catch (Exception e) {
            System.err.println("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.");
            System.exit(1);
        }
    }

    public static String[] buildFFmpegCmd(String input, String output, String videoCodec, String audioCodec,
                                          int crf, String videoBitrate, String audioBitrate, String size, int fps,
                                          String preset, boolean normalize, String subtitles, int pass) {
        List<String> cmd = new ArrayList<>();
        cmd.add("ffmpeg");
        cmd.add("-i");
        cmd.add(input);
        cmd.add("-y");
        cmd.add("-c:v");
        cmd.add(VIDEO_CODECS.getOrDefault(videoCodec, "libx264"));
        cmd.add("-c:a");
        cmd.add(audioCodec != null ? audioCodec : "aac");
        if (crf >= 0) {
            cmd.add("-crf");
            cmd.add(String.valueOf(crf));
        } else if (videoBitrate != null && !videoBitrate.isEmpty()) {
            cmd.add("-b:v");
            cmd.add(videoBitrate);
        }
        if (audioBitrate != null && !audioBitrate.isEmpty()) {
            cmd.add("-b:a");
            cmd.add(audioBitrate);
        }
        if (size != null && !size.isEmpty()) {
            cmd.add("-vf");
            cmd.add("scale=" + size + ":flags=lanczos");
        }
        if (fps > 0) {
            cmd.add("-r");
            cmd.add(String.valueOf(fps));
        }
        if (preset != null && PRESETS.contains(preset)) {
            cmd.add("-preset");
            cmd.add(preset);
        }
        if (normalize) {
            cmd.add("-af");
            cmd.add("loudnorm=I=-16:LRA=11:TP=-1.5");
        }
        if (subtitles != null && !subtitles.isEmpty()) {
            cmd.add("-vf");
            cmd.add("subtitles=" + subtitles);
        }
        if (pass > 0) {
            cmd.add("-pass");
            cmd.add(String.valueOf(pass));
            if (pass == 1) {
                cmd.add("-f");
                cmd.add("null");
                cmd.add("/dev/null");
            }
        }
        cmd.add(output);
        return cmd.toArray(new String[0]);
    }

    public static boolean convertFile(String input, String output, String videoCodec, String audioCodec,
                                      int crf, String videoBitrate, String audioBitrate, String size, int fps,
                                      String preset, boolean twoPass, boolean normalize, String subtitles) {
        if (twoPass && videoBitrate != null && !videoBitrate.isEmpty()) {
            // Первый проход
            String[] pass1Cmd = buildFFmpegCmd(input, "/dev/null", videoCodec, audioCodec, crf, videoBitrate,
                                               audioBitrate, size, fps, preset, normalize, subtitles, 1);
            try {
                Process p1 = Runtime.getRuntime().exec(pass1Cmd);
                if (p1.waitFor() != 0) return false;
            } catch (Exception e) {
                System.err.println("  Ошибка прохода 1: " + e.getMessage());
                return false;
            }
            // Второй проход
            String[] pass2Cmd = buildFFmpegCmd(input, output, videoCodec, audioCodec, crf, videoBitrate,
                                               audioBitrate, size, fps, preset, normalize, subtitles, 2);
            try {
                Process p2 = Runtime.getRuntime().exec(pass2Cmd);
                return p2.waitFor() == 0;
            } catch (Exception e) {
                System.err.println("  Ошибка прохода 2: " + e.getMessage());
                return false;
            }
        } else {
            String[] cmd = buildFFmpegCmd(input, output, videoCodec, audioCodec, crf, videoBitrate,
                                          audioBitrate, size, fps, preset, normalize, subtitles, 0);
            try {
                Process p = Runtime.getRuntime().exec(cmd);
                return p.waitFor() == 0;
            } catch (Exception e) {
                System.err.println("  Ошибка: " + e.getMessage());
                return false;
            }
        }
    }

    public static List<Path> findVideoFiles(String root, boolean recursive) throws IOException {
        List<Path> files = new ArrayList<>();
        Path path = Paths.get(root);
        if (Files.isRegularFile(path)) {
            String ext = getExtension(path.toString()).toLowerCase();
            if (INPUT_EXTS.contains(ext)) {
                files.add(path);
            }
            return files;
        }
        if (!Files.isDirectory(path)) return files;
        if (recursive) {
            Files.walk(path)
                .filter(p -> Files.isRegularFile(p))
                .forEach(p -> {
                    String ext = getExtension(p.toString()).toLowerCase();
                    if (INPUT_EXTS.contains(ext)) files.add(p);
                });
        } else {
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(path)) {
                for (Path p : stream) {
                    if (Files.isRegularFile(p)) {
                        String ext = getExtension(p.toString()).toLowerCase();
                        if (INPUT_EXTS.contains(ext)) files.add(p);
                    }
                }
            }
        }
        return files;
    }

    private static String getExtension(String path) {
        int i = path.lastIndexOf('.');
        return i > 0 ? path.substring(i) : "";
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Использование: java VideoCompress <вход> [--output FILE|DIR] [--codec h264|hevc|vp9|av1] [--audio-codec aac|mp3|opus|ac3] [--crf N] [--video-bitrate N] [--audio-bitrate N] [--size WxH] [--fps N] [--preset PRESET] [--two-pass] [--normalize] [--subtitles FILE] [--recursive] [--threads N] [--overwrite] [--formats]");
            System.exit(1);
        }
        if (args[0].equals("--formats")) {
            System.out.println("Входные форматы: " + INPUT_EXTS);
            System.out.println("Видеокодеки: " + VIDEO_CODECS.keySet());
            System.out.println("Аудиокодеки: " + AUDIO_CODECS);
            System.out.println("Пресеты: " + PRESETS);
            return;
        }

        checkFFmpeg();

        String source = args[0];
        String output = null;
        String videoCodec = "h264";
        String audioCodec = "aac";
        int crf = -1;
        String videoBitrate = null;
        String audioBitrate = null;
        String size = null;
        int fps = 0;
        String preset = "medium";
        boolean twoPass = false;
        boolean normalize = false;
        String subtitles = null;
        boolean recursive = false;
        int threads = 4;
        boolean overwrite = false;

        for (int i=1; i<args.length; i++) {
            if (args[i].equals("--output") && i+1 < args.length) {
                output = args[++i];
            } else if (args[i].equals("--codec") && i+1 < args.length) {
                videoCodec = args[++i];
            } else if (args[i].equals("--audio-codec") && i+1 < args.length) {
                audioCodec = args[++i];
            } else if (args[i].equals("--crf") && i+1 < args.length) {
                crf = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--video-bitrate") && i+1 < args.length) {
                videoBitrate = args[++i];
            } else if (args[i].equals("--audio-bitrate") && i+1 < args.length) {
                audioBitrate = args[++i];
            } else if (args[i].equals("--size") && i+1 < args.length) {
                size = args[++i];
            } else if (args[i].equals("--fps") && i+1 < args.length) {
                fps = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--preset") && i+1 < args.length) {
                preset = args[++i];
            } else if (args[i].equals("--two-pass")) {
                twoPass = true;
            } else if (args[i].equals("--normalize")) {
                normalize = true;
            } else if (args[i].equals("--subtitles") && i+1 < args.length) {
                subtitles = args[++i];
            } else if (args[i].equals("--recursive")) {
                recursive = true;
            } else if (args[i].equals("--threads") && i+1 < args.length) {
                threads = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--overwrite")) {
                overwrite = true;
            }
        }

        if (crf >= 0 && videoBitrate != null) {
            System.err.println("Ошибка: укажите либо --crf, либо --video-bitrate, но не оба.");
            System.exit(1);
        }
        if (!VIDEO_CODECS.containsKey(videoCodec)) {
            System.err.println("Неизвестный видеокодек: " + videoCodec);
            System.exit(1);
        }
        if (!AUDIO_CODECS.contains(audioCodec)) {
            System.err.println("Неизвестный аудиокодек: " + audioCodec);
            System.exit(1);
        }

        List<Path> files = findVideoFiles(source, recursive);
        if (files.isEmpty()) {
            System.out.println("Не найдено видеофайлов в " + source);
            System.exit(1);
        }

        String outDir = output != null && Files.isDirectory(Paths.get(output)) ? output : (output != null ? Paths.get(output).getParent().toString() : "./compressed");
        Files.createDirectories(Paths.get(outDir));
        int total = files.size();
        System.out.println("Найдено " + total + " видеофайлов.");

        ExecutorService executor = Executors.newFixedThreadPool(threads);
        List<Future<String>> futures = new ArrayList<>();
        int[] success = {0};

        for (Path inputFile : files) {
            futures.add(executor.submit(() -> {
                Path rel = Paths.get(source).relativize(inputFile);
                if (rel.toString().isEmpty()) rel = inputFile.getFileName();
                String outPath = Paths.get(outDir, rel.toString().replaceFirst("\\.[^.]+$", "") + ".mp4").toString();
                if (output != null && !Files.isDirectory(Paths.get(output))) {
                    outPath = output;
                }
                if (Files.exists(Paths.get(outPath)) && !overwrite) {
                    return outPath + " уже существует, пропуск.";
                }
                Files.createDirectories(Paths.get(outPath).getParent());
                boolean ok = convertFile(inputFile.toString(), outPath, videoCodec, audioCodec, crf, videoBitrate, audioBitrate,
                                         size, fps, preset, twoPass, normalize, subtitles);
                if (ok) {
                    success[0]++;
                    return "Конвертация " + inputFile + " -> " + outPath;
                }
                return "Ошибка при конвертации " + inputFile;
            }));
        }

        int i = 1;
        for (Future<String> f : futures) {
            System.out.println("[" + i++ + "/" + total + "] " + f.get());
        }
        executor.shutdown();
        System.out.printf("Готово! Успешно: %d, Всего: %d\n", success[0], total);
    }
}

File: video_compress.js
// video_compress.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const { promisify } = require('util');
const yargs = require('yargs');
const { hideBin } = require('yargs/helpers');

const execPromise = promisify(exec);

const SUPPORTED_INPUT = ['.mp4', '.mkv', '.avi', '.mov', '.wmv', '.flv', '.webm', '.m4v', '.3gp', '.ogv'];
const VIDEO_CODECS = { h264: 'libx264', hevc: 'libx265', vp9: 'libvpx-vp9', av1: 'libaom-av1' };
const AUDIO_CODECS = ['aac', 'mp3', 'opus', 'ac3'];
const PRESETS = ['ultrafast', 'superfast', 'veryfast', 'faster', 'fast', 'medium', 'slow', 'slower', 'veryslow'];

function checkFFmpeg() {
    try {
        exec('ffmpeg -version', (err) => {
            if (err) {
                console.error('Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.');
                process.exit(1);
            }
        });
    } catch {
        console.error('Ошибка: ffmpeg не найден.');
        process.exit(1);
    }
}

function buildFFmpegCmd(inputFile, outputFile, videoCodec, audioCodec, crf, videoBitrate, audioBitrate,
                        size, fps, preset, twoPass, normalize, subtitles) {
    const cmd = ['ffmpeg', '-i', inputFile, '-y'];
    cmd.push('-c:v', VIDEO_CODECS[videoCodec] || 'libx264');
    cmd.push('-c:a', audioCodec || 'aac');
    if (crf !== undefined) cmd.push('-crf', String(crf));
    else if (videoBitrate) cmd.push('-b:v', videoBitrate);
    if (audioBitrate) cmd.push('-b:a', audioBitrate);
    if (size) cmd.push('-vf', `scale=${size}:flags=lanczos`);
    if (fps) cmd.push('-r', String(fps));
    if (preset && PRESETS.includes(preset)) cmd.push('-preset', preset);
    if (normalize) cmd.push('-af', 'loudnorm=I=-16:LRA=11:TP=-1.5');
    if (subtitles) cmd.push('-vf', `subtitles=${subtitles}`);
    cmd.push(outputFile);
    return cmd;
}

async function convertFile(inputPath, outputPath, videoCodec, audioCodec, crf, videoBitrate, audioBitrate,
                           size, fps, preset, twoPass, normalize, subtitles) {
    // Двухпроходное кодирование (упрощённо)
    if (twoPass && videoBitrate) {
        const pass1Cmd = buildFFmpegCmd(inputPath, '/dev/null', videoCodec, audioCodec, crf, videoBitrate,
                                        audioBitrate, size, fps, preset, false, normalize, subtitles);
        pass1Cmd.push('-pass', '1', '-f', 'null');
        try {
            await execPromise(pass1Cmd.join(' '));
        } catch (err) {
            console.error(`  Ошибка прохода 1: ${err.stderr}`);
            return false;
        }
        const pass2Cmd = buildFFmpegCmd(inputPath, outputPath, videoCodec, audioCodec, crf, videoBitrate,
                                        audioBitrate, size, fps, preset, false, normalize, subtitles);
        pass2Cmd.push('-pass', '2');
        try {
            await execPromise(pass2Cmd.join(' '));
            return true;
        } catch (err) {
            console.error(`  Ошибка прохода 2: ${err.stderr}`);
            return false;
        }
    } else {
        const cmd = buildFFmpegCmd(inputPath, outputPath, videoCodec, audioCodec, crf, videoBitrate,
                                   audioBitrate, size, fps, preset, false, normalize, subtitles);
        try {
            await execPromise(cmd.join(' '));
            return true;
        } catch (err) {
            console.error(`  Ошибка: ${err.stderr}`);
            return false;
        }
    }
}

function findVideoFiles(root, recursive) {
    const files = [];
    if (fs.existsSync(root) && fs.statSync(root).isFile()) {
        if (SUPPORTED_INPUT.includes(path.extname(root).toLowerCase())) {
            files.push(root);
        }
        return files;
    }
    if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) {
        return files;
    }
    const walk = (dir) => {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);
            if (entry.isDirectory() && recursive) {
                walk(fullPath);
            } else if (entry.isFile()) {
                if (SUPPORTED_INPUT.includes(path.extname(entry.name).toLowerCase())) {
                    files.push(fullPath);
                }
            }
        }
    };
    walk(root);
    return files;
}

async function main() {
    const argv = yargs(hideBin(process.argv))
        .usage('Использование: $0 <вход> [--output FILE|DIR] [--codec h264|hevc|vp9|av1] [--audio-codec aac|mp3|opus|ac3] [--crf N] [--video-bitrate N] [--audio-bitrate N] [--size WxH] [--fps N] [--preset PRESET] [--two-pass] [--normalize] [--subtitles FILE] [--recursive] [--threads N] [--overwrite] [--formats]')
        .demandCommand(1)
        .argv;

    if (argv.formats) {
        console.log('Входные форматы:', SUPPORTED_INPUT.join(', '));
        console.log('Видеокодеки:', Object.keys(VIDEO_CODECS).join(', '));
        console.log('Аудиокодеки:', AUDIO_CODECS.join(', '));
        console.log('Пресеты:', PRESETS.join(', '));
        return;
    }

    checkFFmpeg();

    if (argv.crf !== undefined && argv.videoBitrate) {
        console.error('Ошибка: укажите либо --crf, либо --video-bitrate, но не оба.');
        process.exit(1);
    }

    const source = argv._[0];
    const output = argv.output || null;
    const videoCodec = argv.codec || 'h264';
    const audioCodec = argv.audioCodec || 'aac';
    const crf = argv.crf !== undefined ? argv.crf : null;
    const videoBitrate = argv.videoBitrate || null;
    const audioBitrate = argv.audioBitrate || null;
    const size = argv.size || null;
    const fps = argv.fps || null;
    const preset = argv.preset || 'medium';
    const twoPass = argv.twoPass || false;
    const normalize = argv.normalize || false;
    const subtitles = argv.subtitles || null;
    const recursive = argv.recursive || false;
    const threads = argv.threads || 4;
    const overwrite = argv.overwrite || false;

    const files = findVideoFiles(source, recursive);
    if (files.length === 0) {
        console.log(`Не найдено видеофайлов в ${source}`);
        process.exit(1);
    }

    const outDir = output ? (fs.existsSync(output) && fs.statSync(output).isDirectory() ? output : path.dirname(output)) : './compressed';
    if (!fs.existsSync(outDir)) {
        fs.mkdirSync(outDir, { recursive: true });
    }

    const total = files.length;
    console.log(`Найдено ${total} видеофайлов.`);
    let success = 0;
    const concurrency = Math.min(threads, files.length);
    const queue = [...files];
    const results = [];

    await Promise.all(Array.from({ length: concurrency }, async () => {
        while (queue.length > 0) {
            const inputFile = queue.shift();
            let outFile;
            if (output && !fs.existsSync(output)) {
                outFile = output;
            } else {
                const rel = path.relative(source, inputFile) || path.basename(inputFile);
                outFile = path.join(outDir, path.dirname(rel), path.basename(rel, path.extname(rel)) + '.mp4');
            }
            if (fs.existsSync(outFile) && !overwrite) {
                results.push(`${outFile} уже существует, пропуск.`);
                continue;
            }
            if (!fs.existsSync(path.dirname(outFile))) {
                fs.mkdirSync(path.dirname(outFile), { recursive: true });
            }
            const ok = await convertFile(inputFile, outFile, videoCodec, audioCodec, crf, videoBitrate, audioBitrate,
                                         size, fps, preset, twoPass, normalize, subtitles);
            if (ok) {
                success++;
                results.push(`Конвертация ${inputFile} -> ${outFile}`);
            } else {
                results.push(`Ошибка при конвертации ${inputFile}`);
            }
        }
    }));

    results.forEach((msg, i) => console.log(`[${i+1}/${total}] ${msg}`));
    console.log(`Готово! Успешно: ${success}, Всего: ${total}`);
}

main().catch(console.error);
